python-lib/osisoft_client.py

```diff
@@ -43,6 +43,144 @@ def get_auth(self, auth_type, username, password):
+                                         search_full_hierarchy=search_full_hierarchy, max_count=max_count)
+            counter = 0
+            try:
+                first_item = next(ret)
+                if not previous_item_timestamp or previous_item_timestamp != first_item.get("Timestamp"):
+                    logger.info("Yielding first item")
+                    yield first_item
+                counter += 1
+                for first_item in ret:
+                    yield first_item
+                    counter += 1
+            except Exception as err:
+                if "Error 400" in "{}".format(err):
+                    logger.warning("The time range {} -> {} is too large, splitting the job in two".format(start_date, end_date))
+                    start_timestamp = self.parse_pi_time(start_date)
+                    end_timestamp = self.parse_pi_time(end_date)
+                    new_time_range = (end_timestamp - start_timestamp) / 2
+                    half_time_iso = epoch_to_iso(start_timestamp + new_time_range)
+                    first_half_items = self.get_all_rows_from_item(
+                        pi_tag, data_type, start_date=epoch_to_iso(start_timestamp), end_date=half_time_iso,
+                        interval=interval, sync_time=sync_time, boundary_type=boundary_type, can_raise=True, object_id=object_id,
+                        search_full_hierarchy=search_full_hierarchy, max_count=max_count
+                    )
+                    for item in first_half_items:
+                        yield item
+                    logger.info("Successfuly retrieved first half of {} to {}".format(epoch_to_iso(start_timestamp), half_time_iso))
+                    second_half_items = self.get_all_rows_from_item(
+                        pi_tag, data_type, start_date=half_time_iso, end_date=epoch_to_iso(end_timestamp),
+                        interval=interval, sync_time=sync_time, boundary_type=boundary_type, can_raise=True, object_id=object_id,
+                        search_full_hierarchy=search_full_hierarchy, max_count=max_count
+                    )
+                    for item in second_half_items:
+                        yield item
+                    logger.info("Successfuly retrieved second half of {} to {}".format(half_time_iso, epoch_to_iso(end_timestamp)))
+                else:
+                    logger.error("Error: {}".format(err))
+                    if can_raise:
+                        raise Exception("Error: {}".format(err))
+                    yield {'object_id': "{}".format(object_id), 'Errors': "{}".format(err)}
+            logger.info("Successfuly retrieved time range {} to {}".format(start_date, end_date))
+            if counter == max_count:
+                logger.warning("Number of replies equals maxCount. Shifting startDate and trying one more time.")
+                last_received_timestamp = first_item.get("Timestamp")
```

Comment:
👍 that now read `last_received_timestamp = row.get("Timestamp")` since [8306ad3](https://github.com/dataiku/dss-plugin-osisoft/pull/43/commits/8306ad35215c5cb4c04b9d52f129939f18df258a)